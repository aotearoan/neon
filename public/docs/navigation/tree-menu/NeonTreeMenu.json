{
  "displayName": "NeonTreeMenu",
  "description": "Three level tree menu. This is useful for displaying a hierarchical navigation in NeonSideNav.The top level of the\nmenu is expandable on click but will not navigate to a page. The second level is links to pages and the third level\nis the fragments on that page.",
  "tags": {},
  "exportName": "default",
  "props": [
    {
      "name": "model",
      "description": "The tree model defining the menu.",
      "type": {
        "name": "Array as () => Array<NeonTreeMenuSectionModel>"
      },
      "required": true
    },
    {
      "name": "expandAll",
      "description": "Expand all nodes in the tree, this is useful for providing filtering (e.g. the Neon showcase side navigation menu).",
      "type": {
        "name": "boolean"
      },
      "defaultValue": {
        "func": false,
        "value": "false"
      }
    }
  ],
  "events": [
    {
      "name": "click",
      "description": "Emitted when the user clicks on a menu item",
      "type": {
        "names": [
          "string"
        ]
      }
    }
  ],
  "sourceFiles": [
    "./src/components/navigation/tree-menu/NeonTreeMenu.vue",
    "/Users/andrew/dev/projects/neon/src/components/navigation/tree-menu/NeonTreeMenu.ts"
  ]
}