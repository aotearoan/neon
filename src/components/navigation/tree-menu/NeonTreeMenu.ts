import { defineComponent, ref, watch } from 'vue';
import type { NeonTreeMenuSectionModel } from '@/common/models/NeonTreeMenuSectionModel';
import NeonLink from '../link/NeonLink.vue';
import { useRoute } from 'vue-router';

/**
 * Three level tree menu. This is useful for displaying a hierarchical navigation in NeonSideNav.The top level of the
 * menu is expandable on click but will not navigate to a page. The second level is links to pages and the third level
 * is the fragments on that page.
 */
export default defineComponent({
  name: 'NeonTreeMenu',
  components: {
    NeonLink,
  },
  props: {
    /**
     * The tree model defining the menu.
     */
    model: { type: Array as () => Array<NeonTreeMenuSectionModel>, required: true },
    /**
     * Expand all nodes in the tree, this is useful for providing filtering (e.g. the Neon showcase side navigation menu).
     */
    expandAll: { type: Boolean, default: false },
  },
  emits: [
    /**
     * Emitted when the user clicks on a menu item
     * @type {string} the key of the clicked on menu item.
     */
    'click',
  ],
  setup(props, { emit }) {
    const route = useRoute();
    const url = ref<string | null>(null);

    const click = ($event: KeyboardEvent) => {
      const target = $event.target as HTMLSpanElement;
      if (target.parentElement) {
        target.parentElement.click();
      }
    };

    const onClick = (key: string) => {
      emit('click', key);
    };

    const fragment = (anchor: string) => {
      return anchor.toLowerCase().replace(/\s/g, '-');
    };

    watch(
      () => route.path,
      (value) => (url.value = value),
      { immediate: true },
    );

    return {
      url,
      click,
      onClick,
      fragment,
    };
  },
});
