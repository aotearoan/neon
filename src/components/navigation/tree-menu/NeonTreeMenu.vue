<template>
  <nav v-if="model" class="neon-tree-menu" :class="{ 'neon-tree-menu--expand-all': expandAll }">
    <ul class="no-style">
      <li
        v-for="section in model"
        :key="section.key"
        class="neon-tree-menu__section"
        :class="{
          'neon-tree-menu__section--expanded': section.expanded,
        }"
      >
        <neon-link
          :no-style="true"
          outline-style="none"
          tabindex="-1"
          class="neon-tree-menu__section-link"
          @click="onClick(section.key)"
        >
          <span
            tabindex="0"
            role="link"
            @keydown.space="onClick(section.key)"
            @keypress.space.prevent=""
            class="neon-tree-menu__section-link-label neon-tree-menu__section-link-label--outline-text"
            >{{ section.label }}</span
          >
        </neon-link>
        <ul class="no-style neon-tree-menu__links">
          <li v-for="link in section.children" :key="link.key" class="neon-tree-menu__link-item">
            <neon-link
              :no-style="true"
              outline-style="none"
              tabindex="-1"
              :href="link.href"
              class="neon-tree-menu__link"
            >
              <span
                class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text"
                tabindex="0"
                role="link"
                @keydown.enter="click($event)"
                @keydown.space="click($event)"
                @keypress.space.prevent=""
                >{{ link.label }}</span
              >
            </neon-link>
            <ul
              v-if="link.anchors && link.anchors.length > 0"
              class="no-style neon-tree-menu__anchors"
              :class="{ 'neon-tree-menu__anchors--expanded': expandAll || url === link.href }"
            >
              <li v-for="anchor in link.anchors" :key="anchor">
                <neon-link
                  outline-style="none"
                  tabindex="-1"
                  class="neon-link--no-style neon-tree-menu__anchor"
                  :href="`${link.href}#${fragment(anchor)}`"
                >
                  <span
                    class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text"
                    tabindex="0"
                    role="link"
                    @keydown.space="click($event)"
                    @keydown.enter="click($event)"
                    @keypress.space.prevent=""
                    >{{ anchor }}</span
                  >
                </neon-link>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</template>

<script lang="ts" src="./NeonTreeMenu.ts"></script>
