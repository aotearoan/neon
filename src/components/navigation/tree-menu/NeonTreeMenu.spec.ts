import type { RenderResult } from '@testing-library/vue';
import { fireEvent, render } from '@testing-library/vue';
import NeonTreeMenu from './NeonTreeMenu.vue';
import { router } from '@/../test/unit/test-router';

describe('NeonTreeMenu', () => {
  const model = [
    {
      key: 'feedback',
      label: 'Feedback',
      expanded: false,
      children: [
        {
          key: 'alert',
          label: 'Alert',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
        {
          key: 'note',
          label: 'Note',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
        {
          key: 'notification-counter',
          label: 'Notification Counter',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
      ],
    },
    {
      key: 'navigation',
      label: 'Navigation',
      expanded: true,
      children: [
        {
          key: 'action-menu',
          label: 'Action Menu',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
        {
          key: 'dropdown-menu',
          label: 'Dropdown Menu',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
        {
          key: 'link',
          label: 'Link',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
        {
          key: 'tree-menu',
          label: 'Tree Menu',
          href: '/test',
          anchors: ['Description', 'API', 'Examples'],
        },
      ],
    },
  ];

  let harness: RenderResult;

  beforeEach(() => {
    harness = render(NeonTreeMenu, {
      props: {
        model,
      },
      global: { plugins: [router] },
    });
  });

  it('matches snapshot', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatchSnapshot();
  });

  it('expands all', async () => {
    // given
    const { container, rerender } = harness;
    await rerender({ expandAll: true });
    // when / then
    expect(container.querySelector('.neon-tree-menu--expand-all')).toBeDefined();
    expect(container.querySelectorAll('.neon-tree-menu__anchors--expanded').length).toEqual(7);
  });

  it('emits click event on click section link', async () => {
    // given
    const { container, emitted } = harness;
    // when
    const item = container.querySelectorAll('.neon-tree-menu__section-link').item(0) as HTMLElement;
    await fireEvent.click(item);
    // then
    expect(emitted().click[0]).toEqual(['feedback']);
  });

  it('emits click event on space keydown section link', async () => {
    // given
    const { container, emitted } = harness;
    // when
    const item = container
      .querySelectorAll('.neon-tree-menu__section-link .neon-tree-menu__section-link-label')
      .item(0) as HTMLElement;
    await fireEvent.keyDown(item, { key: 'Space', code: 'Space' });
    // then
    expect(emitted().click[0]).toEqual(['feedback']);
  });
});
