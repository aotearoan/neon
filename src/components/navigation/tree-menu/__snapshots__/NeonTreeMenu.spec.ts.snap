// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonTreeMenu matches snapshot 1`] = `
<nav class="neon-tree-menu">
  <ul class="no-style">
    <li class="neon-tree-menu__section"><a class="neon-link--no-style neon-link--outline-none neon-link neon-tree-menu__section-link neon-tree-menu__section-link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__section-link-label neon-tree-menu__section-link-label--outline-text" role="link" tabindex="0">Feedback</span></span></a>
      <ul class="no-style neon-tree-menu__links">
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Alert</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Note</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Notification Counter</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li class="neon-tree-menu__section--expanded neon-tree-menu__section"><a class="neon-link--no-style neon-link--outline-none neon-link neon-tree-menu__section-link neon-tree-menu__section-link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__section-link-label neon-tree-menu__section-link-label--outline-text" role="link" tabindex="0">Navigation</span></span></a>
      <ul class="no-style neon-tree-menu__links">
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Action Menu</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Dropdown Menu</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Link</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
        <li class="neon-tree-menu__link-item"><a href="/test" class="neon-link--no-style neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-tree-menu__link neon-tree-menu__link" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__link-label neon-tree-menu__link-label--outline-text" role="link" tabindex="0">Tree Menu</span></span></a>
          <ul class="no-style neon-tree-menu__anchors">
            <li><a href="/test#description" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Description</span></span></a></li>
            <li><a href="/test#api" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">API</span></span></a></li>
            <li><a href="/test#examples" class="neon-link--active neon-link--outline-none neon-link neon-link--router-link neon-link--no-style neon-tree-menu__anchor neon-link--no-style neon-tree-menu__anchor" data-testid="link" tabindex="-1"><span class="neon-link__label"><!-- @slot the content of the link --><span class="neon-tree-menu__anchor-label neon-tree-menu__anchor-label--outline-text" role="link" tabindex="0">Examples</span></span></a></li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</nav>
`;
